Add support for `com.hazelcast:hazelcast:5.2.1`

linghengqian opened the pull request:
## What does this PR do?

- Fixes https://github.com/oracle/graalvm-reachability-metadata/issues/32  .
- This is part of https://github.com/oracle/graalvm-reachability-metadata/issues/117 .

## Checklist before merging
- [x] I have properly formatted metadata files (see [CONTRIBUTING](/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md) document)
- [x] I have added thorough tests. (see [this](/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md#Tests))

linghengqian commented on tests/src/com.hazelcast/hazelcast/5.2.1/src/test/java/com_hazelcast/hazelcast/HazelcastTest.java:
```diff
@@ -0,0 +1,168 @@
+import static com.hazelcast.query.Predicates.sql;
+import static org.assertj.core.api.Assertions.assertThat;
+
+class HazelcastTest {
```
- I must point out that I found it completely impossible to complete the nativeTest using `javax.cache:cache-api:1.1.1`, so I did not hang the Jcache unit test into this PR for now. Because it looks like JCache is having the same problem as Log4j2.

linghengqian commented on tests/src/com.hazelcast/hazelcast/5.2.1/src/test/resources/META-INF/native-image/javax.cache/cache-api/1.1.1/serialization-config.json:
```diff
@@ -0,0 +1,18 @@
+{
+  "lambdaCapturingTypes": [],
+  "proxies": [],
+  "types": [
+    {
+      "name": "javax.cache.configuration.FactoryBuilder$SingletonFactory"
+    },
+    {
+      "name": "javax.cache.expiry.AccessedExpiryPolicy"
+    },
+    {
+      "name": "javax.cache.expiry.Duration"
+    },
+    {
+      "name": "javax.cache.expiry.EternalExpiryPolicy"
+    }
+  ]
+}
```
- I think this is acceptable, and I cannot find these serializable json with Conditional Mode by changing `user-code-filter.json`. I had to manually filter `javax.cache:cache-api`'s GraalVM Reactablity Metadata using Standard Agent Mode.

mhalbritter commented on metadata/com.hazelcast/hazelcast/5.2.1/jni-config.json:
```diff
@@ -0,0 +1,166 @@
+    "condition": {
+      "typeReachable": "com.hazelcast.internal.util.JVMUtil"
+    },
+    "name": "com.sun.management.VMOption$Origin",
```
Those `com.sun.management`, `sun.management`, `java.management` etc causes trouble in the hibernate metadata. Could you please remove them from the metadata and try if Hazelcast works without them?

linghengqian replied:
- I submitted a new commit, `./gradlew clean test -Pcoordinates=com.hazelcast:hazelcast:5.2.1` is still normal.
- ![image](https://user-images.githubusercontent.com/20187731/203524482-6948499b-bcdb-4c13-8021-9b1849942bcc.png)

mhalbritter commented on tests/src/com.hazelcast/hazelcast/5.2.1/build.gradle:
```diff
@@ -0,0 +1,44 @@
+}
+
+test {
+    def strings = Arrays.asList("--add-modules", "java.se", "--add-exports", "java.base/jdk.internal.ref=ALL-UNNAMED",
```
Why is this necessary?

mhalbritter replied:
I don't know. If the test still passes, then you can remove it. I guess :)

linghengqian replied:
- Finish.
- ![image](https://user-images.githubusercontent.com/20187731/204861561-84b336a3-81b1-45e3-a9e5-7c8109f623ca.png)

wilkinsona commented:
Like #124, this looks too broad to me. For example, there are many MBean-related entries that may create the same problem as https://github.com/oracle/graalvm-reachability-metadata/pull/113 addressed.

linghengqian commented:
> Like #124, this looks too broad to me. For example, there are many MBean-related entries that may create the same problem as #113 addressed.


- I misunderstood what @mhalbritter meant, I thought he was asking me to keep GraalVM reachability metadata for `javax.management.**`. 
- Looks like `user-code-filter.json` is not able to set restrictions in this regard, I will manually update these two PRs in the near future.

linghengqian commented:
- GraalVM reachability metadata for `com.sun.management.**`, `sun.management.**`, `java.management.**`, `javax.management.**` has been removed.

sdeleuze commented:
Thanks for this PR.

It looks like some resources like `hazelcast-default.xml` and related XSD are missing, and XSD are version specific (Spring Boot 3 uses Hazelcast 5.1 for example). Maybe by default we could ship 5.x schema and not 4.x ones?

@linghengqian Could you please test the refined resources with https://github.com/spring-projects/spring-aot-smoke-tests/tree/main/framework/cache-hazelcast (`../../gradlew build` in this dir) to make sure it works with a typical Spring Boot use case?

linghengqian commented:
@sdeleuze 

- I'll update this PR shortly to test for missing xml and xsd definitions. But is it convenient for you to open the CI of this PR?

- I don't seem to be sure how to put the content of the current PR to https://github.com/spring-projects/spring-aot-smoke-tests for testing, it doesn't seem to directly configure GraalVM Native Build Tools? You mean, I hang the json file of the current PR to `src/main/resources/META-INF/native-image/` folder under https://github.com/spring-projects/spring-aot-smoke-tests/tree/main/framework%2Fcache-hazelcast , and then execute `./gradlew :framework:cache-hazelcast:nativeTest` in the main directory?

linghengqian commented:
@sdeleuze 
- Done. After I updated the unit tests, I also removed the `java.lang.management.**` and `jdk.management.**` related GraalVM reachability metadata. Obviously, it would be great if someone knows how to write a separate PR for nativeTest for `javax.cache:cache-api:1.1.1`. Part of the GraalVM reachability metadata for `javax.cache:cache-api:1.1.1` is located in the `test/resources` folder of the current PR.
-![image](https://user-images.githubusercontent.com/20187731/204989114-2e94b7a8-1cac-4a37-b864-025fa50258c1.png)

- I tested it at https://github.com/spring-projects/spring-aot-smoke-tests/tree/main/framework/cache-hazelcast, but I can't understand its error at all, because the current PR uses `shutdown()` method, and nativeTest works fine. This appears to be a problem with the Spring Framework itself. As an aside, I haven't tested it too deeply.
- ![image](https://user-images.githubusercontent.com/20187731/204991862-086671ab-1ab6-4e14-8802-783cbeb724b8.png)
- ![image](https://user-images.githubusercontent.com/20187731/204992208-07ebbae8-6547-4b94-8685-8a2927f3e3fe.png)
- I am attaching a tarball of the test report generated by Gradle that is not related to this PR. 
- [nativeAppTestReport.zip](https://github.com/oracle/graalvm-reachability-metadata/files/10130214/nativeAppTestReport.zip)

linghengqian commented:
- Can anyone help to restart CI?  It looks like the Github Action device crashed for 6 hours.  Refer to https://github.com/oracle/graalvm-reachability-metadata/actions/runs/3621925163/jobs/6106007671 .

fniephaus commented:
Also, are you sure you invoked the test the same way it is invoked on GitHub actions? I don't see any build output from Native Image, so the generator may not even be running and thus may not cause the build to become stale.

linghengqian commented:
- I may need to investigate more in this PR. All three of my local devices passed `./gradlew clean test -Pcoordinates=com.hazelcast:hazelcast:5.2.1`.

linghengqian commented:
- I realize that the problem with CI has nothing to do with GraalVM Native Build Tools. CI has already caused the Github Actions device to crash due to insufficient memory when executing unit tests under JIT. This is related to the startup of the Hazelcast cluster. 
- I will try again in another branch Structure this PR and resubmit the PR.

linghengqian commented:
- I continue to investigate at https://github.com/oracle/graalvm-reachability-metadata/pull/169 .

fniephaus commented:
So I use [action-tmate](https://github.com/mxschmitt/action-tmate) to log onto a GitHub runner and debug your build and it seems like that some test is waiting for a hazelcast connect (stack trace is below). Maybe you need to start some kind of hazelcast service before the test can run? Also, it may be worth putting a timeout in place for establishing the connect so that you don't have to wait until the GitHub runner times out.

```
"Test worker" #1 prio=5 os_prio=0 cpu=1591.61ms elapsed=82.09s tid=0x00007f1b480273a0 nid=0xd3e waiting on condition  [0x00007f1b4f29c000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
        at java.lang.Thread.sleep(java.base@17.0.5/Native Method)
        at com.hazelcast.client.impl.connection.tcp.WaitStrategy.sleep(WaitStrategy.java:83)
        at com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager.doConnectToCandidateCluster(TcpClientConnectionManager.java:545)
        at com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager.doConnectToCluster(TcpClientConnectionManager.java:428)
        at com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager.connectToCluster(TcpClientConnectionManager.java:389)
        at com.hazelcast.client.impl.clientside.HazelcastClientInstanceImpl.start(HazelcastClientInstanceImpl.java:388)
        at com.hazelcast.client.HazelcastClient.constructHazelcastClient(HazelcastClient.java:472)
        at com.hazelcast.client.HazelcastClient.newHazelcastClientInternal(HazelcastClient.java:420)
        at com.hazelcast.client.HazelcastClient.newHazelcastClient(HazelcastClient.java:119)
        at com_hazelcast.hazelcast.HazelcastTest.testMap(HazelcastTest.java:55)
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@17.0.5/Native Method)
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@17.0.5/NativeMethodAccessorImpl.java:77)
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@17.0.5/DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(java.base@17.0.5/Method.java:568)
        at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
        at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
        at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
        at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
        at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor$$Lambda$191/0x0000000800d10cb0.apply(Unknown Source)
        at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
        at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall$$Lambda$192/0x0000000800d110d0.apply(Unknown Source)
        at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
        at org.junit.jupiter.engine.execution.ExecutableInvoker$$Lambda$343/0x0000000800d39cc0.apply(Unknown Source)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
        at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
        at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor$$Lambda$1664/0x0000000800fa0cf8.execute(Unknown Source)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
...
```

vjovanov commented:
Why didn't we see this exception in the logs?

fniephaus commented:
It's not an exception, it's part of a thread dump. Apparently, it's waiting forever, which is why I suggested using a timeout when establishing the connection.

vjovanov commented:
@dnestoro we should investigate how to kill stuff with SIGQUIT on timeout. This should print the thread dump before exiting.  

I feel this will happen more in the future.

fniephaus commented:
Hi @linghengqian, I don't use Hazelcast so I don't know how to proceed here. At least we now know why the build gets stuck. Feel free to let me know if you run into a similar problem again.
